### PR TITLE
test(multi-collateral): liquidate spot tests

### PR DESCRIPTION
**Pull request by omri-ba-starkware**
> [!NOTE]
> Adds test coverage and helpers for spot-collateral liquidation flows.
> 
> - Extends `perps_tests_facade.cairo` with `LimitOrderInfo`, `create_limit_order`, and `liquidate_spot_asset`, including balance/event validations and support for different `source_position` vs `receive_position`.
> - Adds unit tests in `unit_flow_tests.cairo` covering: liquidation after price drop; multi-step liquidation; exact full-position liquidation; liquidation to empty position; rejection for non-spot and vault-share assets; validation on very small amounts (quote/fee ratio); rejection when position is deleveragable; and liquidation with distinct source/receive positions.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e54a8480674e3517663815c4c6972a6daff63daf. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)

### CODE REVIEW

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> @MohammadNassar1 reviewed 2 files and all commit messages, and made 7 comments.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 7 unresolved discussions (waiting on @ishay-starkware and @omri-ba-starkware).

___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1683 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjphMqS76FcvAc7OCQj:-OjphMqS76FcvAc7OCQk:undefined) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1683)):*
> ```Cairo
>     }
> 
>     fn create_limit_order_with_signature(
> ```





_Suggestion:_
```Cairo
create_limit_order
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1691 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjphPAz5CLTHQxGNnPt:-OjphPAz5CLTHQxGNnPu:bgxjeqr) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1691)):*
> ```Cairo
>         fee_amount: u64,
>         receive_position_id: Option<PositionId>,
>     ) -> (LimitOrder, Signature) {
> ```

I'd suggest creating `LimitOrderInfo` same as we did in OrderInfo.



_Code quote:_
```Cairo
 (LimitOrder, Signature)
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1700 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjphoM5AWIgYeymFwae:-OjphoM5AWIgYeymFwaf:undefined) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1700)):*
> ```Cairo
>         } else {
>             user.position_id
>         };
> ```





_Suggestion:_
```Cairo
        let receive_position_id = match receive_position_id {
            Some(id) => id,
            None => user.position_id,
        };
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1726 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjpiBhs7IVpZL77bOft:-OjpiBhs7IVpZL77bOfu:b6y287y) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1726)):*
> ```Cairo
>         liquidator_order: LimitOrder,
>         liquidator_signature: Signature,
>         liquidator_public_key: PublicKey,
> ```

Why it's needed?



_Code quote:_
```Cairo
liquidator_public_key: PublicKey,
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1730 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjpiVfBC3ofa3KNddFD:-OjpiVfBC3ofa3KNddFE:b-1a8mo) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1730)):*
> ```Cairo
>         actual_amount_base_collateral: i64,
>         actual_liquidator_fee: u64,
>         liquidated_fee_amount: u64,
> ```

Please be consistent with the `liquidate` method.



_Suggestion:_
```Cairo
        liquidated_base: i64,
        liquidated_amount: i64,
        liquidated_insurance_fee: u64,
        liquidator_fee: u64,
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1750 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-Ojpinrx1IFARd805gda:-Ojpinrx1IFARd805gdb:b862rel) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1750)):*
> ```Cairo
> 
>         // Get receive position balance before if different from source
>         let liquidator_receive_balance_before = if liquidator_order
> ```

better to use match here.
Maybe add a flag instead, as you have it in several places.



_Code quote:_
```Cairo
if liquidator_order
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1770 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjpiyPaARA2Zia9O2PG:-OjpiyPaARA2Zia9O2PH:bkd2jhl) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3240962182549d8ab75308763ec1057bc15ad570/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1770)):*
> ```Cairo
>         // Compute the liquidator order hash for event verification (same as contract does)
>         let liquidator_order_hash = liquidator_order
>             .get_message_hash(public_key: liquidator_public_key);
> ```

Should be in `LimitOrderInfo`.



_Code quote:_
```Cairo
        // Compute the liquidator order hash for event verification (same as contract does)
        let liquidator_order_hash = liquidator_order
            .get_message_hash(public_key: liquidator_public_key);
```

**Review by omri-ba-starkware — COMMENTED** [review]
> @omri-ba-starkware made 7 comments and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 5 unresolved discussions (waiting on @ishay-starkware and @MohammadNassar1).

___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1691 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjphPAz5CLTHQxGNnPt:-Oju8OFeCFzK5IGmHwX2:b-896fix) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1691)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

I'd suggest creating `LimitOrderInfo` same as we did in OrderInfo.
</blockquote></details>

Done.
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1726 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjpiBhs7IVpZL77bOft:-Oju9ZA_2Fib8CrYjv-v:b3wm93b) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1726)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Why it's needed?
</blockquote></details>

Once I added LimitOrderInfo no need for public key and signature. Removed.
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1730 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjpiVfBC3ofa3KNddFD:-Oju9xGcBYZRbCZFOfmH:b-qekszu) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1730)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Please be consistent with the `liquidate` method.
</blockquote></details>

This is actually aligned with liquidate_spot. I decided to change liquidate_spot names in order to make it more clear as quote and base terms might be a bit misleading in this flow.
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1750 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-Ojpinrx1IFARd805gda:-OjuEv-V8Bw4S1b9aN28:btr2kmg) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1750)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

better to use match here.
Maybe add a flag instead, as you have it in several places.
</blockquote></details>

Not sure how to use match, I am checking this condition twice so added a flag to save the extra check
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1770 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjpiyPaARA2Zia9O2PG:-OjuCOTz3tIsylsMxYpe:ba3dfob) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1770)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Should be in `LimitOrderInfo`.
</blockquote></details>

Removed
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1683 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjphMqS76FcvAc7OCQj:-Oju6gWoAQ0--LV-Wuk4:bb6un1j) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1683)):*
> ```Cairo
>     }
> 
>     fn create_limit_order_with_signature(
> ```

Done
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1700 at r2](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjphoM5AWIgYeymFwae:-Oju8e7JARafwwwXO2Yc:bb6un1j) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/02aefafd1ecb06c151fbf4be4749159faf9650bf/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1700)):*
> ```Cairo
>         } else {
>             user.position_id
>         };
> ```

Done

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> @MohammadNassar1 reviewed 2 files and all commit messages, made 4 comments, and resolved 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 4 unresolved discussions (waiting on @ishay-starkware and @omri-ba-starkware).

___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1729 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-Ojz6OXS4ErWimH45dER:-Ojz6OXS4ErWimH45dES:bu9lz6n) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3a6cdae5b6a892d862379602a7674fcff9cbc642/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1729)):*
> ```Cairo
>         ref self: PerpsTestsFacade,
>         liquidated_user: User,
>         liquidated_asset_id: AssetId,
> ```

Redundant, can be removed.



_Code quote:_
```Cairo
liquidated_asset_id: AssetId,
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 2962 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzCUEnBYRM_gMOpaJb:-OjzCUEnBYRM_gMOpaJc:b-g1c7v5) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3a6cdae5b6a892d862379602a7674fcff9cbc642/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L2962)):*
> ```Cairo
>     state
>         .facade
>         .validate_total_value(position_id: liquidated_user.position_id, expected_total_value: 200);
> ```

Maybe add another test that makes the position empty. (TV=TR=0)



_Code quote:_
```Cairo
    state
        .facade
        .validate_total_value(position_id: liquidated_user.position_id, expected_total_value: 200);
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 2972 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzCslSEp486eaz5Esg:-OjzCslSEp486eaz5Esh:b-l555al) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3a6cdae5b6a892d862379602a7674fcff9cbc642/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L2972)):*
> ```Cairo
> #[test]
> #[should_panic(expected: 'INVALID_QUOTE_FEE_AMOUNT')]
> fn test_liquidate_spot_very_small_amount() {
> ```

Not sure what this test checks?
Why it expect invalid fee amount? and not not healthier?



_Code quote:_
```Cairo
#[should_panic(expected: 'INVALID_QUOTE_FEE_AMOUNT')]
fn test_liquidate_spot_very_small_amount() {
```
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 3135 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzDEZQ-q_bYcvaE1V5:-OjzDEZQ-q_bYcvaE1V6:bt7qviv) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/3a6cdae5b6a892d862379602a7674fcff9cbc642/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L3135)):*
> ```Cairo
> 
>     // This should FAIL because deleveragable positions have strict validation
>     // that the liquidation cannot satisfy.
> ```

Which strict validation?



_Code quote:_
```Cairo
    // This should FAIL because deleveragable positions have strict validation
    // that the liquidation cannot satisfy.
```

**Review by omri-ba-starkware — COMMENTED** [review]
> @omri-ba-starkware made 4 comments and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 3 unresolved discussions (waiting on @ishay-starkware and @MohammadNassar1).

___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo` line 1729 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-Ojz6OXS4ErWimH45dER:-Ok-6Z-j4Mz2g0858f7y:b-896fix) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/a8541b4234fba41194d39bc976fa6d3869c96e32/workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo#L1729)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Redundant, can be removed.
</blockquote></details>

Done.
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 2962 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzCUEnBYRM_gMOpaJb:-Ok-DKou5lqvAZcbnRRO:bb6un1j) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/a8541b4234fba41194d39bc976fa6d3869c96e32/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L2962)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Maybe add another test that makes the position empty. (TV=TR=0)
</blockquote></details>

Done
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 2972 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzCslSEp486eaz5Esg:-Ok-6o8AAppJ_Cxviwln:b-6saygs) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/a8541b4234fba41194d39bc976fa6d3869c96e32/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L2972)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Not sure what this test checks?
Why it expect invalid fee amount? and not not healthier?
</blockquote></details>

The panic is in validate_trade where we validate for each of the orders (validate_order). We check collateral amount is bigger than the fee amount and if not panic.
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 3135 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzDEZQ-q_bYcvaE1V5:-Ok-AXJcAZhAv5-OgnX4:b-92spmz) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/a8541b4234fba41194d39bc976fa6d3869c96e32/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L3135)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Which strict validation?
</blockquote></details>

Fixed the comment, it fails because TR=0 (before and after) and position is not healthy. Once we implement new TVTR logic this test should actually pass because position is healthier after (but still not healthy)

**Review by MohammadNassar1 — APPROVED** [review]
> @MohammadNassar1 reviewed 2 files and all commit messages, made 2 comments, and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 1 unresolved discussion (waiting on @ishay-starkware and @omri-ba-starkware).

___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 2972 at r3](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-OjzCslSEp486eaz5Esg:-Ok2os9MEhMaVrEdaDze:b-4gor3b) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/a8541b4234fba41194d39bc976fa6d3869c96e32/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L2972)):*
<details><summary><i>Previously, omri-ba-starkware wrote…</i></summary><blockquote>

The panic is in validate_trade where we validate for each of the orders (validate_order). We check collateral amount is bigger than the fee amount and if not panic.
</blockquote></details>

Gotcha.
___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 3157 at r4](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-Ok2p-WSElxVEslooQAu:-Ok2p-WSElxVEslooQAv:b-m0emwj) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/496798c46355daeb462f62e14b1a8040d06283c0/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L3157)):*
> ```Cairo
>             actual_amount_base_collateral: 2,
>             actual_liquidator_fee: 1,
>             liquidated_fee_amount: 100 // Large fee makes it worse!
> ```

Does fee amount = 1 panic? if yes, try to check it.
else check fee =2.



_Suggestion:_
```Cairo
liquidated_fee_amount: 1 // Large fee makes it worse!
```

**Review by omri-ba-starkware — COMMENTED** [review]
> @omri-ba-starkware made 1 comment and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @ishay-starkware and @MohammadNassar1).

___
*[`workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo` line 3157 at r4](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-Ok2p-WSElxVEslooQAu:-Ok3Ntzp1oQ7tW-Sbmpp:b-9qoy28) ([raw file](https://github.com/starkware-libs/starknet-perpetual/blob/496798c46355daeb462f62e14b1a8040d06283c0/workspace/apps/perpetuals/contracts/src/tests/flow_tests/unit_flow_tests.cairo#L3157)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Does fee amount = 1 panic? if yes, try to check it.
else check fee =2.
</blockquote></details>

Done  + added another liquidation with liquidated_fee=2 that panics with the same error

**Review by MohammadNassar1 — APPROVED** [review]
> @MohammadNassar1 reviewed 2 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 1 unresolved discussion (waiting on @ishay-starkware).

**Review by MohammadNassar1 — APPROVED** [review]
:lgtm:

> @MohammadNassar1 made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197#-:-Ok3P5j5DhIstOXeyMHa:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @ishay-starkware and @omri-ba-starkware).

File: workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo [hunk 1]
```diff
@@ -1683,6 +1690,191 @@ pub impl PerpsTestsFacadeImpl of PerpsTestsFacadeTrait {
...
+            quote_amount,
+            fee_asset_id: self.collateral_id,
+            fee_amount,
+            expiration,
+            salt,
+        };
+        let signature = user
+            .account
+            .sign_message(order.get_message_hash(user.account.key_pair.public_key));
+        let hash = order.get_message_hash(user.account.key_pair.public_key);
```
[the remark this answers is not in the document]
Comment (omri-ba-starkware) [reply workspace/apps/perpetuals/contracts/src/tests/flow_tests/perps_tests_facade.cairo]: Fixed

**Review by omri-ba-starkware — COMMENTED** [review]
> @omri-ba-starkware made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-perpetual/197)* status: all files reviewed, 1 unresolved discussion (waiting on @ishay-starkware).
